src/dtx/dtx_resync.c

```diff
@@ -0,0 +1,379 @@
+		if (rc < 0)
+			D_ERROR("Failed to commit the DTX: rc = %d\n", rc);
+		D_FREE(dte);
+	}
+
+	return rc > 0 ? 0 : rc;
+}
+
+static int
+dtx_status_handle(struct dtx_resync_args *dra, bool block)
+{
+	struct ds_cont			*cont = dra->cont;
+	struct pl_obj_layout		*layout = NULL;
+	struct dtx_resync_head		*drh = &dra->tables;
+	struct dtx_resync_entry		*dre;
+	struct dtx_resync_entry		*next;
+	int				 count = 0;
+	int				 err = 0;
+	int				 rc;
+
+	if (drh->drh_count == 0)
+		return 0;
+
+	d_list_for_each_entry_safe(dre, next, &drh->drh_list, dre_link) {
+		int	local_rc;
+
+		if (layout != NULL) {
+			pl_obj_layout_free(layout);
+			layout = NULL;
+		}
+
+		rc = ds_pool_check_leader(dra->po_uuid, &dre->dre_oid,
+					  dra->version, &layout);
+		if (rc <= 0) {
+			if (rc < 0)
+				D_WARN("Not sure about the leader for the DTX "
+				       DF_UOID"/"DF_DTI" (ver = %u): rc = %d, "
+				       "skip it.\n",
+				       DP_UOID(dre->dre_oid),
+				       DP_DTI(&dre->dre_xid), dra->version, rc);
+			else
+				D_DEBUG(DB_TRACE, "Not the leader for the DTX "
+					DF_UOID"/"DF_DTI" (ver = %u) skip it\n",
+					DP_UOID(dre->dre_oid),
+					DP_DTI(&dre->dre_xid), dra->version);
+			dtx_dre_release(drh, dre);
+			continue;
+		}
+
+		rc = dtx_check(dra->po_uuid, cont->sc_uuid,
+			       &dre->dre_dte, layout);
+		if (rc != DTX_ST_COMMITTED && rc != DTX_ST_PREPARED &&
+		    rc != DTX_ST_INIT) {
+			/* We are not sure about whether the DTX can be
+			 * committed or not, then we have to skip it.
+			 */
+			D_WARN("Not sure about whether the DTX "DF_UOID
+			       "/"DF_DTI" can be committed or not: %d (1)\n",
+			       DP_UOID(dre->dre_oid),
+			       DP_DTI(&dre->dre_xid), rc);
+			dtx_dre_release(drh, dre);
+			continue;
+		}
+
+		/* There is CPU yield during DTX operation (check/commit/abort)
+		 * remotely. Then the other DTXs may become committable by race.
+		 * So re-check the remaining DTXs status.
+		 */
+		local_rc = vos_dtx_lookup_cos(cont->sc_hdl, &dre->dre_oid,
+			&dre->dre_xid, dre->dre_hash,
+			dre->dre_intent == DAOS_INTENT_PUNCH ? true : false);
+		if (local_rc == 0) {
+			/* The DTX is in CoS cache (committable), do nothing. */
+			dtx_dre_release(drh, dre);
+			continue;
+		}
+
+		/* The DTX is not in CoS cache, but it has been committed on
+		 * some remote replica(s), then let's commit the it globally.
+		 */
+		if (rc == DTX_ST_COMMITTED)
+			goto commit;
+
+		if (local_rc != -DER_NONEXIST) {
+			D_WARN("Not sure about whether the DTX "DF_UOID
+			       "/"DF_DTI" can be committed or not: %d (2)\n",
+			       DP_UOID(dre->dre_oid),
+			       DP_DTI(&dre->dre_xid), rc);
+			dtx_dre_release(drh, dre);
+			continue;
+		}
+
+		/* It is possible that other ULT has committed the DTX during
+		 * we handle other DTXs, then such DTX will not be in the CoS
+		 * cache. Let's re-check the on-disk DTX table.
+		 *
+		 * Set the @dkey_hash parameter as zero, then it will skip CoS
+		 * cache since we just did that in above check.
+		 */
```

Comment:
Not sure why it can't be done in same vos_dtx_check_committable() call. Actually, I was wondering why can't we just call one function to get transaction status instead of calling three functions (dtx_check, vos_dtx_lookup_cos and vos_dtx_check_commitable)?

Reply:
We load the DTX status from local disk firstly, then check remote replica status, there will be CPU yield for the remote DTX status check and it is possible that some others will change the DTX status during such CPU yield.

Reply:
I mean instead of calling three functions, we may introduce one function and do a single call to get dtx status.

Reply:
minor, the "dkey_hash" usage seems a little tricky, maybe use another special flag ("skip_cos"?) is more readable